src/traceViewer/web/ui/filmStripView.tsx

```diff
@@ -0,0 +1,183 @@
+      const ratio = Math.min(1, (measure.width - 20) / idealWidth);
+      const image = new Image((idealWidth * ratio) | 0, (idealHeight * ratio) | 0);
+      image.src = imageURL(context, previewVideo.fileName, previewIndex);
+      return new Promise<HTMLImageElement>(f => image.onload = () => f(image));
+    }
+    loadPreviewImage().then(image => setPreviewImage(image));
+  }, [previewVideo, previewMetaInfo, previewIndex, measure.width]);
+
+  return <div className='film-strip' ref={ref}>{
+    Array.from(metaInfos.entries()).map(([video, metaInfo]) => <FilmStripLane
+      context={context}
+      boundaries={boundaries}
+      video={video}
+      metaInfo={metaInfo}
+      width={measure.width}
+    />)
+  }
+  {preview && previewMetaInfo && previewImage &&
+    <div className='film-strip-hover' style={{
+      width: previewImage.width + 'px',
+      height: previewImage.height + 'px',
+      top: measure.bottom + 5 + 'px',
+      left: Math.min(preview.clientX, measure.width - previewImage.width - 10) + 'px',
+    }}>
+      <img src={previewImage.src} width={previewImage.width} height={previewImage.height} />
+    </div>
+  }
+  </div>;
+}
+
+const FilmStripLane: React.FunctionComponent<{
+  context: ContextEntry,
+  boundaries: Boundaries,
+  video: PageVideoTraceEvent,
+  metaInfo: MetaInfo,
+  width: number,
+}> = ({ context, boundaries, video, metaInfo, width }) => {
+  // Position clip on timeline. Note: should be aligned with css.
+  const paddingLeft = 20 /* timeline zero */;
+  const frameWidth = 80;
+  const frameMargin = 5;
+  const gapLeft = (metaInfo.startTime - boundaries.minimum) / (boundaries.maximum - boundaries.minimum) * width;
+  const gapRight = (boundaries.maximum - metaInfo.endTime) / (boundaries.maximum - boundaries.minimum) * width;
+  const effectiveWidth = (metaInfo.endTime - metaInfo.startTime) / (boundaries.maximum - boundaries.minimum) * (width - paddingLeft);
+
+  const frameCount = effectiveWidth / (frameWidth + frameMargin) | 0;
+  const frameStep = metaInfo.frames / frameCount;
+  const frameHeight = frameWidth / metaInfo.width * metaInfo.height | 0;
+  const frameGap = frameCount <= 1 ? 0 : (effectiveWidth - (frameWidth + frameMargin) * frameCount) / (frameCount - 1);
+
+  const videoId = context.created.contextId + ':' + video.fileName;
+
+  const frames: JSX.Element[] = [];
+  for (let i = 0; i < metaInfo.frames; i += frameStep) {
+    let index = i | 0;
+    // Always show last frame.
+    if (Math.floor(i + frameStep) >= metaInfo.frames)
+      index = metaInfo.frames - 1;
+    frames.push(<div className='film-strip-frame' key={videoId + ':' + i} style={{
+      width: frameWidth + 'px',
+      height: frameHeight + 'px',
+      backgroundImage: `url(${imageURL(context, video.fileName, index)})`,
+      backgroundSize: `${frameWidth}px ${frameHeight}px`,
+      marginRight: (frameMargin / 2 + frameGap) + 'px',
+    }} />);
+  }
+
+  return <div className='film-strip-lane' key={videoId} style={{
```

Comment:
Generally I'd let the outside component do the keying, but I guess this is fine because you have a unique id

Reply:
Ack.